Support mixed protocol LBs

kiall opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR adds support for configuring a service of type LoadBalancer with more than 1 protocol. For example, this allows configuring a LoadBalancer with both TCP and UDP port 53 (for a DNS server), or TCP and UDP 443 for HTTPS + QUIC (HTTP 3.0).

**Which issue(s) this PR fixes**:

Fixes #23880

**Special notes for your reviewer**:

Mixed protocol LBs are supported by Azure and MetalLB. Other providers MAY support this, however, I'm unable to verify the implementations and have opted to reject mixed protocol LBs on these CPIs where necessary (this turned out to only need rejection in the GCP provider)

An example service to enable and use this feature:

    apiVersion: v1
    kind: Service
    metadata:
      name: mixed-protocol
    spec:
      type: LoadBalancer
      ports:
        - name: dns-udp
          port: 53
          protocol: UDP
        - name: dns-tcp
          port: 53
          protocol: TCP
      selector:
        app: my-dns-server

If the feature gate is enabled, while a CPI without support is enabled, the following event will be recorded:

	 Type     Reason                      Age                   From                Message
	 ----     ------                      ----                  ----                -------
	 Warning  CreatingLoadBalancerFailed  112s (x71 over 32m)   service-controller  Error creating load balancer (will retry): failed to ensure load balancer for service default/mixed-protocol: Cannot create an GCP load balancer with mixed protocols

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Added support for mixed protocol services of type=LoadBalancer, enabled through the use of a new `ServiceLoadBalancerMixedProtocol` feature gate.
```

grahamhayes commented:
/ok-to-test

itsc0lin commented:
would welcome support for this!

mattkeeler commented:
It would be great to have support for mixed protocol LBs.

andyzhangx commented:
@kiall any progress on this PR?

alvaroaleman commented on pkg/apis/core/annotation_key_constants.go:
```diff
@@ -83,6 +83,13 @@ const (
+	// It should be a string with either "true", or "false".
+	//
+	// Not all cloud providers support this annotation, though Azure, OpenStack and MetalLB do.
+	AnnotationLoadBalancerAllowMixedProtocolsKey = "service.beta.kubernetes.io/load-balancer-allow-mixed-protocols"
```
Can you expand a little on the reasoning for an annotation? A check if a Service contains ports for more than one protocol shouldn't be too hard to write and could be used as a basis for having providers that do not support that emit an event. Or are there more reasons for having it?

kiall replied:
Really, I think the annotation should be little more than a stopgap measure to handle compatibility and allow people (i.e. CPI vendors) time to update - I gave some detail in the PR message above.

feiskyer commented:
>  I decided to make this feature opt-in via the annotation, however I believe that is the wrong approach long term.

For new features, it's preferable to add a feature gate instead of annotations. For the first release, it may be alpha so that users can try it. And in later beta releases, enable it by default.

Miouge1 commented:
@kiall looks good, but a couple of merge conflicts shows up. Maybe a rebase is needed?

kiall committed:
Support mixed protocol LBs

Adds a new Feature Gate which, enabled, allows mixed protocol Services of type
LoadBalancer to be created. Not all CPI implementations support this - Azure,
and MetalLB have been tested to work. Some providers (e.g. AWS) do support
mixed protocols, however the CPI implementation does not - these will need to
be updated by the respective CPI authors. Finally, other providers (e.g. GCE)
do not support mixed protocols at all currently.

Issue #23880

kiall commented:
@feiskyer I've updated to make use of a feature gate instead, I believe your right - that's a cleaner and more obvious solution.

kiall commented:
/retest

kiall commented:
Hey @dims / @anguslees - if either of you get a chance to review - Thanks! :)

Let me know if you believe it should be assigned elsewhere.

dims commented:
/uncc

andrewsykim commented:
This change looks like it needs a [KEP](https://github.com/kubernetes/enhancements/tree/master/keps#kubernetes-enhancement-proposals-keps).

skwokie commented:
Hi, I'd like to use this feature when it becomes available on Mac (docker desktop) and EKS, and I'd like ask about the status of it? Thanks.

prestonvanloon commented:
Any update? Would really like to use mixed protocols

alvaroaleman commented:
@kiall Do you currently have the capacity to continue with this? Else I'd like to go ahead and write a KEP for the validation changes to make sure we manage to get this in within the 1.17 cycle

Hyzhou commented:
Any update? Some game developer deploy their application used mixed protocols. when use the k8s to deploy it, mixed protocols was not supported in service.
If not to merge this PR. If we have another way to support this feature.

adampl commented:
Does it really need a KEP and a feature gate? Looks more like a bugfix to me.

dcbw commented:
Tim says it was originally disabled because there were possibilities for "being charged twice" depending on your cloud provider. Also Tim wants it to be opt-in.

We need a KEP for this first to explore the issues and proposed solutions.

adampl commented:
But "being charged twice" is possible only if someone enabled this in their config, so what's the problem? And who cares that some cloud provider can't do their job right?

DreamRivulet commented:
Any update on this? without this, we have to create several service for each port, which is inconvenient.

xiaoxubeii commented:
Any updates? We also really need mixed protocal.

jrx-sjg commented:
Oh my god, can anyone solve the conflicts? We really need mixed protocols for our SIP infrastructure.

janosi commented:
The KEP pull request is on its way to kubernetes/enhancements. Please check and comment.

prestonvanloon commented:
/remove-lifecycle rotten

alvaroaleman commented:
@prestonvanloon I don't think there is much point in keeping this PR open, there is a KEP for this, discussion should continue there: https://github.com/kubernetes/enhancements/pull/1438 or the issue at https://github.com/kubernetes/kubernetes/issues/23880

prestonvanloon commented:
OK thanks. I hadn't seen those links yet.

>

cmluciano commented:
Let's continue discussion in #23880 since there is now a merged KEP added

/close
